File: 25.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* newHead = new ListNode(-1);
        newHead->next = head;
        ListNode* last = newHead; 
        ListNode* start = head;
        
        int cnt = 0;
        ListNode* t = head;
        while(t){
            t = t->next;
            cnt++;
            if(cnt==k){
               ListNode* pre = start;
               ListNode* cur = start->next;
               while(cur && cur!=t){
                   ListNode * next = cur->next;
                   cur->next = pre;
                   pre = cur;
                   cur = next;
               }
               last->next = pre;
               last = start;
               start->next = t;
               start = t;
               cnt=0;
            }
        }
        return newHead->next;
    }
};

File: jianzhi-offer/25/25.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
	var (
		head = &ListNode{}
		cur  = head
	)
	for l1 != nil || l2 != nil {
		if l1 == nil {
			cur.Next = l2
			break
		}

		if l2 == nil {
			cur.Next = l1
			break
		}

		if l1.Val < l2.Val {
			cur.Next = l1
			l1 = l1.Next
		} else {
			cur.Next = l2
			l2 = l2.Next
		}

		cur = cur.Next
	}

	return head.Next
}
